index.html

```diff
@@ -417,42 +384,57 @@
     };
 
     callback ThingEventListener = void (ConsumedThing thing);
+
+    enum ConsumedThingEvents {
+        "propertychanged",
+        "propertyremoved",
+        "propertyadded",
+        "actionchanged"
+    };
   </pre>
   </section>
 
   <section> <h2>The Thing Server API</h2>
   <pre class="idl">
     interface ExposedThing {
+
+      // define TD modifiers
       ExposedThing addProperty(ThingProperty property);
+      ExposedThing removeProperty(ThingProperty property);
+
       ExposedThing addAction(ThingAction action);
+      ExposedThing removeAction(ThingAction action);
+
       ExposedThing addEvent(ThingEvent event);
+      ExposedThing removeEvent(ThingEvent event);
 
-      Promise&lt;void&gt; register();
+      Promise&lt;void&gt; register(optional USVString directory);
       Promise&lt;void&gt; unregister();
 
       Promise&lt;void&gt; start();
       Promise&lt;void&gt; stop();
 
       Promise&lt;void&gt; emitEvent(DOMString eventName, any payload);
 
-      // define request handlers
-      ExposedThing onPropertyRetrieve(PropertyRetrieveHandler handler);
-      ExposedThing onPropertyUpdate(PropertyHandler handler);
+      // define request handlers (one per request type, so no events here)
+      ExposedThing onRetrieveProperty(PropertyHandler handler);
+      ExposedThing onAddProperty(PropertyHandler handler);
+      ExposedThing onRemoveProperty(PropertyHandler handler);
 
-      ExposedThing onActionInvocation(ActionHandler handler);
+      ExposedThing onInvokeAction(ActionHandler handler);
 
-      ExposedThing onObserve(DOMString event, ObserveHandler handler);
-      ExposedThing onUnobserve(DOMString event, UnobserveHandler handler);
+      ExposedThing onAddEventHandler(DOMString event, ObserveHandler handler);
+      ExposedThing onRemoveEventHandler(DOMString event, ObserveHandler handler);
 
-      any onSemanticTranslate(Property property, SemanticType type);
+      ExposedThing onAddSemanticTranslator(SemanticTranslator handler);
     };
 
     ExposedThing implements ConsumedThing;
 
-    callback PropertyRetrieveHandler = void (ThingProperty property, USVString url);
-    callback PropertyUpdateHandler = void (ThingProperty property, USVString url);
-    callback ActionHandler = void (ThingAction action, USVString url);
-    callback ObserveHandler = void (DOMString eventName, USVString url);
+    callback PropertyHandler = void (ThingProperty property, USVString from);
+    callback ActionHandler = void (ThingAction action, USVString from);
```

Comment:
Oh, Ok. I interpreted it as (3) setting the callback that gets called when the action is invoked.